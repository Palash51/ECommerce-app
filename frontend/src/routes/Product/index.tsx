import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {RootState} from 'typesafe-actions';
import { History, LocationState } from "history";
import styled from 'styled-components';

import { detailsProduct } from '../../reducers/products/actions'
import { Link } from 'react-router-dom';


interface IProps {
    match: {
        params: {
            id: string
        }
    }
    history: History<LocationState>;
}

const Wrapper = styled.div`

`;

const BackLinkWrapper = styled.div`
    padding: 1rem;
`;

const ProductDetailWrapper = styled.div`
    display: flex;
    padding: 1rem;
    justify-content: space-between;
    align-items: flex-start;
    flex-wrap: wrap;
`;

const ProductImageWrapper = styled.div`
    flex: 1 1 30rem;
    > img{
        max-width: 60rem;
        width: 100%;
    }
`;

const ProductInfoWrapper = styled.div`
    flex: 1 1 30rem;
    > ul{
        list-style-type: none;
        padding: 0;
        > li{
            margin-bottom: 1rem;
        }
    }
    
`;

const ProductActionWrapper = styled.div`
    flex: 1 1 30rem;
    border: 0.1rem #808080 solid;
    border-radius: 0.5rem;
    background-color: #f8f8f8;
    padding: 1rem;
    > ul{
        list-style-type: none;
        padding: 0;
        > li{
            margin-bottom: 1rem;
        }
    }
`;

const CartButtonWrapper = styled.button`
    padding: 1rem;
    border: 0.1rem #808080 solid;
    border-radius: 0.5rem;
    cursor: pointer;
    &:hover {
        border: 0.1rem #404040 solid;
    }
    background-color: #f0c040;

`;


const Product = (props:IProps) => {
    const [qty, setQty] =  useState(1);
    const productDetails =  useSelector((state:RootState) => state.productDetails);
    const { product, loading, error} = productDetails;
    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(detailsProduct(props.match.params.id));
        return () => {
            //
        }
    }, [])

    const updateQuantity = (e: React.ChangeEvent<HTMLSelectElement>) => {
        setQty(Number(e.target.value))
    }

    const handleAddToCart = () => {
        props.history.push("/cart/" + props.match.params.id + "?qty=" + qty)
    }

    return (
    
    <Wrapper>
        <BackLinkWrapper>
            <Link to="/">Back to result</Link>
        </BackLinkWrapper>
    {loading ? <div>loading...</div> : 
    error ? <div>{error}</div> : 
    (
        <ProductDetailWrapper>
            <ProductImageWrapper>
                <img src={product?.image} alt="product" ></img>
            </ProductImageWrapper>
            <ProductInfoWrapper>
                <ul data-testid="product-details">
                    <li>
                        <h4>{product?.name}</h4>
                    </li>
                    <li>
                        {product?.rating} Stars ({product?.numReviews} Reviews)
                    </li>
                    <li>
                        Price: <b>${product?.price}</b>
                    </li>
                    <li>
                        Description:
                        <div>
                            {/* {product.description} */}
                        </div>
                    </li>

                </ul>
            </ProductInfoWrapper>
            <ProductActionWrapper>
                    <ul>
                        <li>
                            Price: {product?.price}
                        </li>
                        <li>
                            Status: {product?.countInStock > 0 ? "In Stock" : "Out of stock"}
                        </li>
                        <li>
                            {/* TODO: Need to remove downlevelIteration */}
                            Quantity: <select value={qty} onChange={(e) => updateQuantity(e)} >
                            {[...Array(product?.countInStock).keys()].map((x) => (
                            <option key={x + 1} value={x + 1}>
                                {x + 1}
                            </option>
                            ))}
                            </select>
                        </li>
                        
                        <li style={{ display:'flex', flexDirection: 'column'}}>
                        {product?.countInStock > 0 &&  <CartButtonWrapper onClick={handleAddToCart}>Add to cart</CartButtonWrapper>}
                        </li>
                        
                    </ul>
            </ProductActionWrapper>
        </ProductDetailWrapper>
        )}
    </Wrapper>
    );
}

export default Product;