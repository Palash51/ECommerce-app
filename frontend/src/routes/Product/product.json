{
    "brand": "Nike",
    "category": "Shirts",
    "countInStock": 6,
    "image": "/images/d1.jpg",
    "name": "Slim Shirts",
    "numReviews": 10,
    "price": 60,
    "rating": 4.5,
    "_id": "1"
}
