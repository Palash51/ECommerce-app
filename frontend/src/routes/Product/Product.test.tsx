import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store'
import { createMemoryHistory } from 'history'
import { MemoryRouter } from 'react-router-dom';

import { mount, shallow } from 'enzyme';
import thunk from 'redux-thunk'
import Product from './index';
import product from './product.json'


const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));


describe('Product detail component', () => {

    it('Should match snapshot', () => {
        const store = mockStore({   
            productDetails: product
        })
        const history = createMemoryHistory('/product/1')
        const productProps = {
            match: {
                params: {
                    id: "1"
                }
            },
        }
        const wrapper = mount(
            <Provider store={store}>
              <BrowserRouter>
              <MemoryRouter>
              <Product  {...productProps} history={history}/>
              </MemoryRouter>
            </BrowserRouter>
            </Provider>,
        )
        expect(wrapper).toMatchSnapshot();
    })

    it('Test products details on detail page', () => {
        const store = mockStore({   
            productDetails: product
        })
        const history = createMemoryHistory('/product/1')
        const productProps = {
            match: {
                params: {
                    id: "1"
                }
            },
        }
        const wrapper = mount(
            <Provider store={store}>
              <BrowserRouter>
              <MemoryRouter>
              <Product  {...productProps} history={history}/>
              </MemoryRouter>
            </BrowserRouter>
            </Provider>,
        )
        expect(wrapper.find('[data-testid="product-details"]').children()).toHaveLength(4);
    })    
})

