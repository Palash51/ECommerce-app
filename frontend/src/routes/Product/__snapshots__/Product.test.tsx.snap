// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product detail component Should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <MemoryRouter>
        <Router
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "cznb66",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "cznb66",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
        >
          <Product
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "key": "iru481",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "iru481",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            match={
              Object {
                "params": Object {
                  "id": "1",
                },
              }
            }
          >
            <styled.div>
              <div
                className="sc-bdnylx dEMiuQ"
              >
                <styled.div>
                  <div
                    className="sc-gtssRu eNMFFu"
                  >
                    <Link
                      to="/"
                    >
                      <LinkAnchor
                        href="/"
                        navigate={[Function]}
                      >
                        <a
                          href="/"
                          onClick={[Function]}
                        >
                          Back to result
                        </a>
                      </LinkAnchor>
                    </Link>
                  </div>
                </styled.div>
                <styled.div>
                  <div
                    className="sc-dlnjPT uqAsM"
                  >
                    <styled.div>
                      <div
                        className="sc-hKFyIo laFNfS"
                      >
                        <img
                          alt="product"
                        />
                      </div>
                    </styled.div>
                    <styled.div>
                      <div
                        className="sc-eCApGN ddGgNw"
                      >
                        <ul
                          data-testid="product-details"
                        >
                          <li>
                            <h4 />
                          </li>
                          <li>
                             Stars (
                             Reviews)
                          </li>
                          <li>
                            Price: 
                            <b>
                              $
                            </b>
                          </li>
                          <li>
                            Description:
                            <div />
                          </li>
                        </ul>
                      </div>
                    </styled.div>
                    <styled.div>
                      <div
                        className="sc-jSFkmK ezrcbM"
                      >
                        <ul>
                          <li>
                            Price: 
                          </li>
                          <li>
                            Status: 
                            Out of stock
                          </li>
                          <li>
                            Quantity: 
                            <select
                              onChange={[Function]}
                              value={1}
                            >
                              <option
                                key="1"
                                value={1}
                              >
                                1
                              </option>
                            </select>
                          </li>
                          <li
                            style={
                              Object {
                                "display": "flex",
                                "flexDirection": "column",
                              }
                            }
                          />
                        </ul>
                      </div>
                    </styled.div>
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </Product>
        </Router>
      </MemoryRouter>
    </Router>
  </BrowserRouter>
</Provider>
`;
